vsc/baseline/dns/students.py

```diff
@@ -0,0 +1,229 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from einops import rearrange
+from .layers import Attention, NetVLAD, BinarizationLayer
+from .similarities import ChamferSimilarity, VideoComparator
+
+
+model_urls = {
+    "dns_cg_student": "https://mever.iti.gr/distill-and-select/models/dns_cg_student.pth",
+    "dns_fg_att_student": "https://mever.iti.gr/distill-and-select/models/dns_fg_att_student.pth",
+    "dns_fg_bin_student": "https://mever.iti.gr/distill-and-select/models/dns_fg_bin_student.pth",
+}
+
+
+def check_dims(features, mask=None, axis=0):
+    if features.ndim == 4:
+        return features, mask
+    elif features.ndim == 3:
+        features = features.unsqueeze(axis)
+        if mask is not None:
+            mask = mask.unsqueeze(axis)
+        return features, mask
+    else:
+        raise Exception(
+            "Wrong shape of input video tensor. The shape of the tensor must be either "
+            "[N, T, R, D] or [T, R, D], where N is the batch size, T the number of frames, "
+            "R the number of regions and D number of dimensions. "
+            "Input video tensor has shape {}".format(features.shape)
+        )
+
+
+class CoarseGrainedStudent(nn.Module):
+    def __init__(
+        self,
+        dims=512,
+        attention=True,
+        transformer=True,
+        transformer_heads=8,
+        transformer_feedforward_dims=2048,
+        transformer_layers=1,
+        netvlad=True,
+        netvlad_clusters=64,
+        netvlad_outdims=1024,
+        pretrained=False,
+        **kwargs
+    ):
+        super().__init__()
+        self.student_type = "cg"
+
+        if attention:
+            self.attention = Attention(dims, norm=False)
+        if transformer:
+            encoder_layer = nn.TransformerEncoderLayer(
+                dims, transformer_heads, transformer_feedforward_dims, batch_first=True
+            )
+            self.transformer = nn.TransformerEncoder(
+                encoder_layer, transformer_layers, nn.LayerNorm(dims)
+            )
+
+        if netvlad:
+            self.netvlad = NetVLAD(dims, netvlad_clusters, outdims=netvlad_outdims)
+
+        if pretrained:
+            self.load_state_dict(
+                torch.hub.load_state_dict_from_url(model_urls["dns_cg_student"])[
```

Comment:
Just a note -- candidates' code has to work without internet access. Our baselines don't necessarily have to conform to that, but it might be good to demonstrate that.

I guess the challenge here is that DnS requires a few models, and we don't have structured configs where we could specify several of them without a lot of plumbing.

I don't want to make this unnecessarily messy or difficult, so maybe we keep it as is.

Reply:
This is resolved for DnS if we go with the torchscripts solution. 
But it still holds for DINO. So if this is not a big issue, I suggest leaving it as is.